File: backend/app/services/page_text.py
"""Join/split full-document text that contains `=== PAGE N ===` markers (1-based page numbers)."""
import re
from typing import Optional

_PAGE_HDR = re.compile(r"^=== PAGE (\d+) ===\s*$", re.MULTILINE)


def join_marked_pages(parts: list[tuple[int, str]]) -> str:
    """``parts`` are (page_index 0-based, body). Produces a single string for Text IQ / summaries."""
    blocks = []
    for idx, body in sorted(parts, key=lambda x: x[0]):
        blocks.append(f"=== PAGE {idx + 1} ===\n{(body or '').strip()}")
    return "\n\n".join(blocks)


def split_marked_pages(full: str) -> dict[int, str]:
    """Returns ``page_index`` (0-based) -> body. If no markers, entire string is page 0."""
    if not full or not full.strip():
        return {}
    matches = list(_PAGE_HDR.finditer(full))
    if not matches:
        return {0: full.strip()}
    out: dict[int, str] = {}
    for i, m in enumerate(matches):
        page_num = int(m.group(1))
        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(full)
        body = full[start:end].strip()
        out[page_num - 1] = body
    return out


def merge_split_into_optional_base(split: dict[int, str], n_pages: int) -> list[str]:
    """Return list of length ``n_pages`` with text per index, default empty string."""
    return [split.get(i, "").strip() for i in range(n_pages)]

File: frontend/src/lib/pageText.ts
/** Mirrors backend `page_text.join_marked_pages` / `split_marked_pages` (markers are 1-based). */

import type { DocumentPageT, DocumentT } from "./api";

export function joinMarkedPages(parts: [pageIndex: number, body: string][]): string {
  const blocks: string[] = [];
  for (const [idx, body] of [...parts].sort((a, b) => a[0] - b[0])) {
    blocks.push(`=== PAGE ${idx + 1} ===\n${(body ?? "").trim()}`);
  }
  return blocks.join("\n\n");
}

/** True if `full` uses `=== PAGE N ===` markers (multi-page joined OCR). */
export function documentHasPageMarkers(full: string | null | undefined): boolean {
  if (!full?.trim()) return false;
  return /^=== PAGE \d+ ===\s*$/m.test(full);
}

export function splitMarkedPages(full: string): Record<number, string> {
  if (!full?.trim()) return {};
  const re = /^=== PAGE (\d+) ===\s*$/gm;
  const matches = [...full.matchAll(re)];
  if (matches.length === 0) return { 0: full.trim() };
  const out: Record<number, string> = {};
  for (let i = 0; i < matches.length; i++) {
    const pageNum = parseInt(matches[i][1]!, 10);
    const start = matches[i].index! + matches[i][0].length;
    const end = i + 1 < matches.length ? matches[i + 1].index! : full.length;
    out[pageNum - 1] = full.slice(start, end).trim();
  }
  return out;
}

/** Source English text per page for Lingua — matches backend `_page_source_for_translate`. */
export function pageSourceForTranslate(doc: DocumentT, page: DocumentPageT): string {
  const a = (page.corrected_page_abstract || page.page_abstract || "").trim();
  if (a) return a;
  const o = (page.corrected_ocr_text || page.ocr_text || "").trim();
  if (o) return o;
  for (const field of [doc.corrected_abstract, doc.abstract] as const) {
    if (!field?.trim()) continue;
    const sp = splitMarkedPages(field);
    const t = (sp[page.page_index] || "").trim();
    if (t) return t;
  }
  for (const field of [doc.corrected_ocr, doc.raw_ocr] as const) {
    if (!field?.trim()) continue;
    const sp = splitMarkedPages(field);
    const t = (sp[page.page_index] || "").trim();
    if (t) return t;
  }
  return "";
}
